fleetctl journal fails on OSX

I have a fleet cluster running on Vagrant, and my host machine is running OSX. When trying to run `fleetctl journal hello.service` from the host, it panics and fails. Logs below.

```
$ echo $FLEETCTL_TUNNEL
127.0.0.1:2222
$ fleetctl list-units
UNIT        LOAD    ACTIVE  SUB DESC        MACHINE
hello.service   loaded  active  running Hello World fb25a864.../172.12.8.101
$ fleetctl journal hello.service
panic: open /proc/sys/kernel/random/boot_id: no such file or directory

goroutine 1 [running]:
runtime.panic(0x319180, 0xc2100ec870)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/runtime/panic.c:266 +0xb6
github.com/coreos/fleet/machine.readLocalBootID(0x533df, 0x3ac860)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/machine/state.go:54 +0x88
github.com/coreos/fleet/machine.IsLocalMachineState(0xc2100ab040, 0x3ac860)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/machine/state.go:48 +0x1e
main.journalAction(0xc210085c40)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/fleetctl/journal.go:59 +0x421
github.com/coreos/fleet/third_party/github.com/codegangsta/cli.Command.Run(0x3adb00, 0x7, 0x0, 0x0, 0x422cd0, ...)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/github.com/codegangsta/cli/command.go:73 +0x994
github.com/coreos/fleet/third_party/github.com/codegangsta/cli.(*App).Run(0xc21007d000, 0xc21000a000, 0x3, 0x3, 0x95d158, ...)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/github.com/codegangsta/cli/app.go:111 +0x855
main.main()
    /Users/bacongobbler/go/src/github.com/coreos/fleet/fleetctl/cmd.go:137 +0xd98

goroutine 3 [chan receive]:
github.com/coreos/fleet/third_party/github.com/golang/glog.(*loggingT).flushDaemon(0x7ea760)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/github.com/golang/glog/glog.go:839 +0x50
created by github.com/coreos/fleet/third_party/github.com/golang/glog.init·1
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/github.com/golang/glog/glog.go:406 +0x276

goroutine 4 [syscall]:
runtime.goexit()
    /usr/local/Cellar/go/1.2/libexec/src/pkg/runtime/proc.c:1394

goroutine 13 [chan receive]:
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/agent.func·002()
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/agent/forward.go:38 +0x45
created by github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/agent.ForwardToAgent
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/agent/forward.go:49 +0x150

goroutine 7 [IO wait]:
net.runtime_pollWait(0x9587b8, 0x72, 0x0)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/runtime/netpoll.goc:116 +0x6a
net.(*pollDesc).Wait(0xc210076fb0, 0x72, 0x957098, 0x23)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/net/fd_poll_runtime.go:81 +0x34
net.(*pollDesc).WaitRead(0xc210076fb0, 0x23, 0x957098)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/net/fd_poll_runtime.go:86 +0x30
net.(*netFD).Read(0xc210076f50, 0xc210080000, 0x1000, 0x1000, 0x0, ...)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/net/fd_unix.go:204 +0x2a0
net.(*conn).Read(0xc210000348, 0xc210080000, 0x1000, 0x1000, 0x9afba0, ...)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/net/net.go:122 +0xc5
bufio.(*Reader).fill(0xc21004e480)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/bufio/bufio.go:91 +0x110
bufio.(*Reader).Read(0xc21004e480, 0xc21004fca0, 0x5, 0x5, 0x5, ...)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/bufio/bufio.go:159 +0x1a4
io.ReadAtLeast(0x958db0, 0xc21004e480, 0xc21004fca0, 0x5, 0x5, ...)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/io/io.go:288 +0xf6
io.ReadFull(0x958db0, 0xc21004e480, 0xc21004fca0, 0x5, 0x5, ...)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/io/io.go:306 +0x71
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*streamPacketCipher).readPacket(0xc21004fc80, 0xc200000009, 0x958db0, 0xc21004e480, 0x1fb, ...)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/cipher.go:143 +0xa4
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*connectionState).readPacket(0xc210077240, 0xc21004e480, 0x10db0, 0xc2100e2d80, 0xc2100820e8, ...)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/transport.go:110 +0x74
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*transport).readPacket(0xc210077240, 0xc210082600, 0x1fb, 0x0, 0x0, ...)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/transport.go:106 +0x3a
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*handshakeTransport).readOnePacket(0xc2100520f0, 0xc210082000, 0xc210082600, 0x1fb, 0x1fb, ...)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/handshake.go:158 +0x97
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*handshakeTransport).readLoop(0xc2100520f0)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/handshake.go:138 +0x27
created by github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.newClientTransport
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/handshake.go:101 +0xee

goroutine 8 [chan receive]:
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*handshakeTransport).readPacket(0xc2100520f0, 0xc210082600, 0x1fb, 0x1fb, 0x0, ...)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/handshake.go:129 +0x34
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*mux).onePacket(0xc21007e9a0, 0x0, 0x0)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/mux.go:220 +0x3a
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*mux).loop(0xc21007e9a0)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/mux.go:195 +0x3e
created by github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.newMux
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/mux.go:124 +0x118

goroutine 9 [chan receive]:
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*Client).handleGlobalRequests(0xc21003fdc0, 0xc21004a540)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/client.go:130 +0x3a
created by github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.NewClient
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/client.go:53 +0xbb

goroutine 10 [chan receive]:
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*Client).handleChannelOpens(0xc21003fdc0, 0xc2100786e0)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/client.go:139 +0x45
created by github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.NewClient
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/client.go:54 +0xdf

goroutine 11 [semacquire]:
sync.runtime_Syncsemacquire(0xc21003fd90)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/runtime/sema.goc:257 +0xca
sync.(*Cond).Wait(0xc21003fd80)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/sync/cond.go:62 +0x89
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*mux).Wait(0xc21007e9a0, 0x0, 0x0)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/mux.go:109 +0x8d
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.func·003()
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/client.go:56 +0x37
created by github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.NewClient
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/client.go:58 +0x10c

goroutine 12 [chan receive]:
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*forwardList).handleChannels(0xc21003fdd0, 0xc210078840)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/tcpip.go:157 +0x45
created by github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.NewClient
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/client.go:59 +0x15e

goroutine 15 [chan receive]:
github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.DiscardRequests(0xc21004a700)
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/connection.go:81 +0x3a
created by github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh.(*Client).dial
    /Users/bacongobbler/go/src/github.com/coreos/fleet/third_party/code.google.com/p/gosshnew/ssh/tcpip.go:328 +0x162

goroutine 16 [runnable]:
net/http.(*persistConn).readLoop(0xc21004f600)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/net/http/transport.go:778 +0x68f
created by net/http.(*Transport).dialConn
    /usr/local/Cellar/go/1.2/libexec/src/pkg/net/http/transport.go:528 +0x607

goroutine 17 [select]:
net/http.(*persistConn).writeLoop(0xc21004f600)
    /usr/local/Cellar/go/1.2/libexec/src/pkg/net/http/transport.go:791 +0x271
created by net/http.(*Transport).dialConn
    /usr/local/Cellar/go/1.2/libexec/src/pkg/net/http/transport.go:529 +0x61e
```

--- a/machine/state.go
+++ b/machine/state.go
@@ -3,6 +3,7 @@ package machine
 import (
 	"io/ioutil"
 	"net"
+	"path/filepath"
 	"strings"
 
 	log "github.com/coreos/fleet/third_party/github.com/golang/glog"
@@ -27,7 +28,7 @@ type MachineState struct {
 // NewDynamicMachineState generates a MachineState object with
 // the values read from the local system
 func CurrentState() MachineState {
-	bootID := readLocalBootID()
+	bootID := readLocalBootID("/")
 	publicIP := getLocalIP()
 	return MachineState{BootID: bootID, PublicIP: publicIP, Metadata: make(map[string]string, 0)}
 }
@@ -45,13 +46,14 @@ func (s MachineState) MatchBootID(bootID string) bool {
 
 // IsLocalMachineState checks whether machine state matches the state of local machine
 func IsLocalMachineState(ms *MachineState) bool {
-	return ms.BootID == readLocalBootID()
+	return ms.BootID == readLocalBootID("/")
 }
 
-func readLocalBootID() string {
-	id, err := ioutil.ReadFile(bootIDPath)
+func readLocalBootID(root string) string {
+	fullPath := filepath.Join(root, bootIDPath)
+	id, err := ioutil.ReadFile(fullPath)
 	if err != nil {
-		panic(err)
+		return ""
 	}
 	return strings.TrimSpace(string(id))
 }

--- a/machine/state_test.go
+++ b/machine/state_test.go
@@ -1,6 +1,9 @@
 package machine
 
 import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
 	"testing"
 )
 
@@ -107,5 +110,40 @@ func TestStateMatchBootID(t *testing.T) {
 		if ok := tt.m.MatchBootID(tt.s); !ok {
 			t.Errorf("#%d: expected %v", i, true)
 		}
+	}
+}
+
+func TestReadLocalBootIDMissing(t *testing.T) {
+	dir, err := ioutil.TempDir(os.TempDir(), "fleet-")
+	if err != nil {
+		t.Fatalf("Failed creating tempdir: %v", err)
+	}
+	defer os.RemoveAll(dir)
+
+	if bootID := readLocalBootID(dir); bootID != "" {
+		t.Fatalf("Received incorrect bootID: %s", bootID)
+	}
+}
+
+func TestReadLocalBootIDFound(t *testing.T) {
+	dir, err := ioutil.TempDir(os.TempDir(), "fleet-")
+	if err != nil {
+		t.Fatalf("Failed creating tempdir: %v", err)
+	}
+	defer os.RemoveAll(dir)
+
+	tmpBootIDPath := filepath.Join(dir, "/proc/sys/kernel/random/boot_id")
+	err = os.MkdirAll(filepath.Dir(tmpBootIDPath), os.FileMode(0755))
+	if err != nil {
+		t.Fatalf("Failed setting up fake boot ID path: %v", err)
+	}
+
+	err = ioutil.WriteFile(tmpBootIDPath, []byte("pingpong"), os.FileMode(0644))
+	if err != nil {
+		t.Fatalf("Failed writing fake boot ID file: %v", err)
+	}
+
+	if bootID := readLocalBootID(dir); bootID != "pingpong" {
+		t.Fatalf("Received incorrect bootID %q, expected 'pingpong'", bootID)
 	}
 }

